hr_attendance_report_theoretical_time/models/hr_leave.py

```diff
@@ -0,0 +1,45 @@
+# Copyright 2017-2019 Tecnativa - Pedro M. Baeza
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models
+
+
+class HrLeave(models.Model):
+    _inherit = "hr.leave"
+
+    def _create_resource_leave(self):
+        """On leave creation, trigger the recomputation of the involved
+        records."""
+        res = super()._create_resource_leave()
+        self._check_theoretical_hours()
+        return res
+
+    def _remove_resource_leave(self):
+        """On leave cancellation, trigger the recomputation of the involved
+        records."""
+        res = super()._remove_resource_leave()
+        self._check_theoretical_hours()
+        return res
+
+    def _check_theoretical_hours(self):
+        """Recomputes all the theoretical hours that corresponds to the
+        interval of dates and employee of the leaves.
+
+        :param: self: Leave recordset.
+        """
+        to_recompute = self.env["hr.attendance"]
+        for record in self.filtered(lambda x: x.date_from and x.date_to):
+            from_datetime = record.date_from.replace(
+                hour=0, minute=0, second=0, microsecond=0
+            )
+            to_datetime = record.date_to.replace(
+                hour=23, minute=59, second=59, microsecond=99999
+            )
+            to_recompute |= self.env["hr.attendance"].search(
+                [
+                    ("employee_id", "=", record.employee_id.id),
+                    ("check_in", ">=", from_datetime),
+                    ("check_in", "<=", to_datetime),
```

Comment:
the comparation has been done by two args with the same datatype `datetime`, and tz is UTC as default

Reply:
@dzungtran89 
For example we have Record (hr.attendance, 1) : check in (GMT+7) = 2023-04-10 05:00:00 -> convert to UTC (to stored in database) = 2023-04-09 22:00:00.

Date = 2023-04-10
=> from_datetime = 2023-04-10 00:00:00
=> to_datetime = 2023-04-10 23:59:59
==> Record (hr.attendance, 1) will not be searched by domain
```
[
("check_in", ">=", "2023-04-10 00:00:00"),
("check_in", "<=", "2023-04-10 23:59:59"),
]
```

Reply:
> Date = 2023-04-10
Is this value extracted from `date_from` which is defined as `datetime`? https://github.com/odoo/odoo/blob/16.0/addons/hr_holidays/models/hr_leave.py#L175

Reply:
Yes it is, if i create a time off for 2023-04-10, my timezone is GTM+7, my working schedule from 08:00 to 17:00
date_from will be stored db: `2023-04-10 08:00:00`
=> from_datetime = 2023-04-10 00:00:00
=> to_datetime = 2023-04-10 23:59:59
Because i'm off on 2023-04-10, so i want my attendance on that date must be take into account. But as i explain above, Record (hr.attendance, 1) will be missed

Reply:
The below is my test
- leave from/to is the same with your case
- the attendance check in is at `2023-4-10 8:15`
```
>>> self.env['hr.leave'].browse(14)
hr.leave(14,)
>>> self.env['hr.leave'].browse(14).date_from
datetime.datetime(2023, 4, 10, 1, 0)
>>> self.env['hr.leave'].browse(14).date_to
datetime.datetime(2023, 4, 10, 10, 0)
>>> from_datetime = self.env['hr.leave'].browse(14).date_from.replace(hour=0, minute=0, second=0,microsecond=0)
>>> from_datetime
datetime.datetime(2023, 4, 10, 0, 0)
>>> to_datetime = self.env['hr.leave'].browse(14).date_to.replace(hour=23, minute=59, second=39,microsecond=99999)
>>> to_datetime
datetime.datetime(2023, 4, 10, 23, 59, 39, 99999)
>>> self.env['hr.attendance'].search([('employee_id', '=', 1), ('check_in', '>=', from_datetime), ('check_in', '<=', to_datetime)]
... )
hr.attendance(30,)
>>> self.env['hr.attendance'].search([('employee_id', '=', 1), ('check_in', '>=', from_datetime), ('check_in', '<=', to_datetime)]).mapped('check_in
')
[datetime.datetime(2023, 4, 10, 1, 15, 37)]
```

Reply:
You should test the attendance which checked in before 7am, for example 2023-4-10 6:15

Reply:
thx for the comments, the fix is coming soon

Reply:
after having a double check, I think there would probably be a comparation between `orange` and `apple` in this context. In order to simplify the compare, let's check based on the user tz first

- if the tz is GMT+7, timeoff is on 2023-04-10, and working schedule is is from 8am to 5pm. And you have an attendance at "2023-4-10 6:15", so the search's output will be None. There would be an output if you replace the 8am to 0h, and 5pm to 24h of the given date. Does it make sense?

Then, in the context of the code
- Both of `record.date_from` and `record.date_to` have already been at `utc` tz. By replacing to 0h and 24h repectively, the tz here is still utc, unless we specific the tz here at step of recent replace

Feel free to correct me if I miss something here, or we can continue discussing via direct message

Reply:
For me, the code is not wrong because both `record.date_from` and `record.date_to` have already been at `utc tz` but it is not good when we call `replace(hour=0, minute=0, second=0, microsecond=0)`. There will be more records than what we need.
@nguyenminhchien  `For example we have Record (hr.attendance, 1) : check in (GMT+7) = 2023-04-10 05:00:00 -> convert to UTC (to store in database) = 2023-04-09 22:00:00.`

in this case, I think `record.date_from` will be ` 2023-04-09 22:00:00` and `record.date_from.eplace(hour=0, minute=0, second=0, microsecond=0)` will be ` 2023-04-09 00:00:00` -> System will compute more records than we need but it is not wrong

Please correct me if I'm wrong

Reply:
@lehoangan thanks for the comment, are you okay if I would raise this issue to the author when creating the PR to oca?

Reply:
Yes, maybe we can ask why the author wants to replace hour=0